### PR TITLE
Homework 2 / Tests

### PR DESCRIPTION
**Pull request by lumenalux**
# Homework 2 / Tests
## Unit tests
- Improved the quality of unit tests
- Added more cases
## Integration tests
- Added integration tests for controller
- Added integration tests for subscription service
## E2E tests
- Created a test environment for E2E tests
- Added github workflow for E2E testing
## Mutation tests
- Added github workflow for mutation testing in `.github/workflows/.golangci.yml`

### CODE REVIEW

**Review by kachan1208 — COMMENTED** [review]
Do I get it correctly that you haven't gotten any review from a partner?

File: internal/subscription/service.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+func (s *Service) Subscribe(email string) error {
+	subscribed, err := s.IsSubscribed(email)
+	if err != nil {
+		return err
+	}
+	if subscribed {
+		return ErrAlreadySubscribed
+	}
+
+	return s.Storage.Append([]string{email})
```
Comment (delveper) [inline internal/subscription/service.go:31]: Is't it nicer to have this signature?:
```go
Append(record ...string) error
```

File: pkg/storage/csv_test.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package storage
+
+import (
+	"os"
+	"reflect"
+	"testing"
+)
+
+func TestCSVStorage(t *testing.T) {
```
Comment (delveper) [inline pkg/storage/csv_test.go:9]: Splitting this test to check methods separately could be a good idea. You can also do complex integration testing, though.

File: docker-compose.e2e.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    build:
+      context: .
+      dockerfile: DockerfileE2E
+    volumes:
+      - storage_volume:/app/storage/
+      - ./test/E2E/config:/app/configs
+    ports:
+      - "8080:8080"
+    depends_on:
+      - smtp
```
Comment (delveper) [inline docker-compose.e2e.yml:13]: Is it possible that the SMTP service have not started before the test runs?

File: .github/workflows/e2e-tests.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    - name: Docker Compose Up
+      run: docker-compose -f docker-compose.e2e.yml up -d
+
+    - name: Install Newman
+      run: sudo npm install -g newman
+
+    - name: Run E2E Tests
+      run: |
+        sleep 30
```
Comment (delveper) [inline .github/workflows/e2e-tests.yml:32]: What if application takes longer than 30 seconds to start? It looks like a hack anyways.

File: internal/rate/kuna_provider.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+	return resp, nil
+}
+
+func (p *KunaProvider) extractRateFromResponse(resp *http.Response) (float32, error) {
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return p.config.DefaultRate, err
```
Comment (delveper) [inline internal/rate/kuna_provider.go:62]: It might be a good idea to add more detailed error messages that indicate what exactly went wrong.

File: internal/rate/kuna_provider_test.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+func TestKunaProviderExchangeRate(t *testing.T) {
+	tests := []struct {
+		name           string
+		mockHTTPClient *StubHTTPClient
+		expectedRate   float32
+		expectedError  string
+	}{
+		{
+			name: "Success",
```
Comment (delveper) [inline internal/rate/kuna_provider_test.go:22]: You could add a test case for when the response body JSON structure is correct but the exchange rate is not a float64.

**Review by delveper — COMMENTED** [review]
The codebase is well-organized and adheres to best practices. However, it could be more consistent and improved by adding more comprehensive tests, especially for edge cases and error conditions, and by providing more detailed error messages to help with debugging.

File: README.md [hunk 7]
```diff
@@ -14,18 +14,12 @@ This is an API that provides the current exchange rate between Bitcoin and the U
...
-   ```
-
 ## Usage
 
-1. Run the Docker container:
+1. Up the docker compose:
 
    ```bash
-   docker run -p 8080:8080 gses2-app
+   docker-compose up --build
```
Comment (kachan1208) [inline README.md:22]: -d to run in background

**Review by kachan1208 — COMMENTED** [review]
reviewed partially

File: docker-compose.e2e.yml (hunk 3 above)
Comment (lumenalux) [reply docker-compose.e2e.yml:13]: This is a guarantee that without fake smtp and kuna_api images the image of the main application will not start. But yes, it does not give a 100% guarantee that the applications themselves will start before the application itself starts. That's why I added a delay of 30 second

```bash
sleep 30
```
On the other hand this delay it is a hack, and I will fix it

File: internal/rate/kuna_provider_test.go (hunk 6 above)
Comment (lumenalux) [reply internal/rate/kuna_provider_test.go:22]: Fixed in 41836c8

File: docker-compose.e2e.yml (hunk 3 above)
Comment (lumenalux) [reply docker-compose.e2e.yml:13]: Removed `sleep 30` in fd09753

File: test/integration/subscription_integratin_test.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+		t.Fatalf("failed to create temporary file: %v", err)
+	}
+
+	defer os.Remove(tmpFile.Name())
+
+	csvStorage := storage.NewCSVStorage(tmpFile.Name())
+
+	service := subscription.NewService(csvStorage)
+
+	t.Run("Subscribe a new email", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply test/integration/subscription_integratin_test.go]: Added in 457e17a

File: test/integration/subscription_integratin_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
+// Subscription integration contains integration tests for the subscription layer,
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply test/integration/subscription_integratin_test.go]: Fixed typo in  7421632

File: test/integration/controller_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+		&email.StubSMTPClientFactory{Client: &email.StubSMTPClient{}, Err: nil},
+	)
+
+	if err != nil {
+		t.Fatalf("error creating email sender service: %v", err)
+	}
+
+	rateService := rate.NewService(&rate.StubProvider{Rate: 42})
+
+	tests := []struct {
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply test/integration/controller_integration_test.go:57]: Added more test cases in a226e38

File: pkg/config/config_test.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+			}
+
+			if config != testData.expectedConfig {
+				t.Errorf("Unexpected configuration. Got: %+v, Expected: %+v", config, testData.expectedConfig)
+			}
+		})
+	}
+}
+
+func TestLoadNonExistentFile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply pkg/config/config_test.go]: Combined in b86fd0d

File: internal/subscription/service.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package subscription
+
+import (
+	"errors"
+)
+
+type Storage interface {
+	Append(record []string) error
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply internal/subscription/service.go]: Changed in 451ecbe

File: README.md (hunk 7 above)
Comment (lumenalux) [reply README.md]: Added `--deatach` parameter in e7b8f8e

File: cmd/gses2-app/main.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	"gses2-app/internal/subscription"
+	"gses2-app/internal/transport"
+	"gses2-app/pkg/config"
+	"gses2-app/pkg/storage"
+)
+
+func main() {
+	config, err := config.Load("./configs/config.yaml")
+	if err != nil {
+		log.Fatalf("Error, config wasn't loaded: %s", err)
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply cmd/gses2-app/main.go]: Fixed in 25be05b

File: cmd/gses2-app/main.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+	rateService, emailSubscriptionService, emailSenderService, err := createServices(&config)
+	controller := controllers.NewAppController(
+		rateService,
+		emailSubscriptionService,
+		emailSenderService,
+	)
+
+	if err != nil {
+		log.Fatalf("Connection error: %s", err)
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply cmd/gses2-app/main.go]: Fixed in 25be05b

File: cmd/gses2-app/main.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+	return mux
+}
+
+func startServer(port string, handler http.Handler) {
+	fmt.Printf("Starting server on port %s\n", port)
+
+	err := http.ListenAndServe(fmt.Sprintf(":%s", port), handler)
+	if err != nil {
+		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply cmd/gses2-app/main.go:81]: Fixed in 25be05b

File: cmd/gses2-app/main.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+	router := transport.NewHTTPRouter(controller)
+
+	mux := http.NewServeMux()
+	router.RegisterRoutes(mux)
+
+	return mux
+}
+
+func startServer(port string, handler http.Handler) {
+	fmt.Printf("Starting server on port %s\n", port)
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply cmd/gses2-app/main.go]: Changed in e8ee6a1

File: cmd/gses2-app/main.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+	"gses2-app/internal/email"
+	"gses2-app/internal/rate"
+	"gses2-app/internal/subscription"
+	"gses2-app/internal/transport"
+	"gses2-app/pkg/config"
+	"gses2-app/pkg/storage"
+)
+
+func main() {
+	config, err := config.Load("./configs/config.yaml")
```
[the remark this answers is not in the document]
Comment (kachan1208) [reply cmd/gses2-app/main.go:20]: prefer to use envs.

File: internal/controllers/controller.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+	if json.NewEncoder(w).Encode(exchangeRate) != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func (ac *AppController) SubscribeEmail(w http.ResponseWriter, r *http.Request) {
+	email := r.FormValue("email")
+	err := ac.EmailSubscriptionService.Subscribe(email)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusConflict)
```
Comment (kachan1208) [inline internal/controllers/controller.go:56]: here you can't say which type of error it was. 
prefer to divide errors into different levels for public api.

in cases when your subscription service fails you need to return 500
if it's just a problem that such email already exists you need to return different code.

File: internal/controllers/controller_test.go [hunk 19]
```diff
@@ -0,0 +1,327 @@
+package controllers
```
Comment (kachan1208) [inline internal/controllers/controller_test.go:1]: like the way you organised stubs. easy to use and clear. 

the only change I prefer to do in such cases is to specify configuration of stub in stub constructor, rest of things are great.

File: internal/controllers/controller_test.go [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+	return nil, errors.New("Get subscriptions error")
+}
+
+type StubEmailSenderServiceError struct{}
+
+func (m *StubEmailSenderServiceError) SendExchangeRate(
+	rate float32,
+	subscribers []string,
+) error {
+	return errors.New("Send email error")
```
Comment (kachan1208) [inline internal/controllers/controller_test.go:77]: prefer errors in lower case

File: internal/controllers/controller_test.go [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+func TestGetRate(t *testing.T) {
```
Comment (kachan1208) [inline internal/controllers/controller_test.go:80]: it looks like you wanted to have table tests but you only have one case.
don't do such things for "future" purpose only as it may never be needed.

Pls combine tests into single table test or remove table tests.

File: internal/controllers/controller_test.go [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+			)
+
+			req, _ := http.NewRequest("GET", "/rate", nil)
+
+			rr := httptest.NewRecorder()
+
+			controller.GetRate(rr, req)
+
+			if status := rr.Code; status != tt.expectedStatus {
+				t.Errorf("GetRate returned wrong status code: got %v, expected %v", status, tt.expectedStatus)
```
Comment (kachan1208) [inline internal/controllers/controller_test.go:112]: almost nobody uses std testing functionality.
prefer to use github.com/testify/require as it will make your tests much simpler.

File: internal/controllers/controller_test.go (hunk 19 above)
Comment (kachan1208) [inline internal/controllers/controller_test.go:1]: Additional reading on table tests and how they should be done: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

File: internal/email/message.go [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+
+type EmailMessage struct {
+	from    string
+	to      []string
+	subject string
+	body    string
+}
+
+func NewEmailMessage(
+	config config.EmailConfig,
```
Comment (kachan1208) [inline internal/email/message.go:20]: according to DDD architecture, config may be defined in the package where it's used.

File: internal/email/message.go [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+	return &EmailMessage{
+		from:    config.From,
+		to:      to,
+		subject: config.Subject,
+		body:    body.String(),
+	}, nil
+}
+
+func (e *EmailMessage) Prepare() []byte {
+	message := fmt.Sprintf("From: %s\r\n"+
```
Comment (kachan1208) [inline internal/email/message.go:44]: what about golang templates?

File: internal/email/send.go [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+	"io"
+)
+
+var (
+	errNoRecepients = errors.New("no recepiets")
+)
+
+type MailClient interface {
+	Mail(string) error
+	Rcpt(string) error
```
Comment (kachan1208) [inline internal/email/send.go:14]: what does this name mean?

File: cmd/gses2-app/main.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lumenalux) [reply cmd/gses2-app/main.go:20]: That's what I want to do in the future. But this is too much work for now, because the application was originally heavily tied to the configuration file. I will have to rewrite some tests, change a little the work with testing and docker compose, rewrite `README.md` and `README_ua.md`.  So far, I don't have my _**HW-3**_ assignment ready yet. As soon as I finish it, I will definitely create a branch with transition from `config.yaml` to `env`

File: internal/email/send.go [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		return err
+	}
+
+	_, err = writer.Write(message)
+	if err != nil {
+		return err
+	}
+
+	err = writer.Close()
```
Comment (kachan1208) [inline internal/email/send.go:48]:
```suggestion
	return writer.Close()
```

File: internal/email/service.go [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+	dialer TLSConnectionDialer,
+	factory SMTPClientFactory,
+) (*SenderService, error) {
+	client := NewSMTPClient(config.SMTP, dialer, factory)
+	clientConnection, err := client.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	SMTPClient := &SenderService{
```
Comment (kachan1208) [inline internal/email/service.go:25]: move this to return statement, you don't need variable here.

File: internal/email/service.go [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+		config:     config,
+		connection: clientConnection,
+	}
+
+	return SMTPClient, nil
+}
+
+type TemplateData struct {
```
Comment (kachan1208) [inline internal/email/service.go:33]: Go template?

File: internal/email/service.go [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+	emailAddresses []string,
+) error {
+
+	templateData := TemplateData{Rate: fmt.Sprintf("%.2f", exchangeRate)}
+	email, err := NewEmailMessage(es.config.Email, emailAddresses, templateData)
+	if err != nil {
+		return err
+	}
+
+	err = SendEmail(es.connection, email)
```
Comment (kachan1208) [inline internal/email/service.go:48]:
```suggestion
	return SendEmail(es.connection, email)
```

File: internal/email/service_test.go [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+			expectedErr: errFactoryError,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			config := &config.Config{}
+			service, err := NewSenderService(config, tt.dialer, tt.factory)
+
+			unexpectedErr := err != nil && tt.expectedErr == nil
```
Comment (kachan1208) [inline internal/email/service_test.go:58]: rewrite next 3 lines with usage of testify/require

File: pkg/config/config.go [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	ErrReadFile      = errors.New("failed to read config file")
+	ErrUnmarshalYAML = errors.New("failed to unmarshal yaml data")
+)
+
+func Load(filename string) (Config, error) {
```
Comment (kachan1208) [inline pkg/config/config.go:16]: prefer to use envs instead of file config. 

12factor app, config section.

File: internal/email/send.go (hunk 26 above)
Comment (delveper) [reply internal/email/send.go:48]: Returning `nil` as the happy path in the last line is idiomatic and more explicit approach.

Comment (kachan1208) [reply internal/email/send.go:48]: https://github.com/golang/go/blob/master/src/net/http/client.go#L457

https://github.com/golang/go/blob/master/src/net/http/client.go#L488

https://github.com/golang/go/blob/master/src/net/http/client.go#L488

...

And I can continue until eternity)
The prettiness of Go is that you can read it and get the approaches.

Comment (delveper) [reply internal/email/send.go:48]: Please don't. You won! 😄 Maybe it is overkill but I like idea of being explicit for all costs. It reduces cognitive load from reviewer.

File: internal/email/service.go (hunk 28 above)
Comment (lumenalux) [reply internal/email/service.go:33]: Yes, I use it for the go template

File: pkg/config/config.go (hunk 31 above)
Comment (lumenalux) [reply pkg/config/config.go:16]: I will completely switch to `env` variables in HW3. This is quite a big piece of work and would be good as part of a refactoring homework assignment

File: internal/controllers/controller.go (hunk 18 above)
Comment (lumenalux) [reply internal/controllers/controller.go:59]: Divided in 29b26da9664f77ded4e5cc21df5f2b1723c7b08b

File: internal/controllers/controller_test.go (hunk 19 above)
Comment (lumenalux) [reply internal/controllers/controller_test.go:1]: Fixed in 2e6a4d14e8142e6bff6794ba8cd971f023559033

File: internal/controllers/controller_test.go (hunk 20 above)
Comment (lumenalux) [reply internal/controllers/controller_test.go]: Removed in 2e6a4d14e8142e6bff6794ba8cd971f023559033

File: internal/controllers/controller_test.go (hunk 21 above)
Comment (lumenalux) [reply internal/controllers/controller_test.go:62]: Combined in 2e6a4d14e8142e6bff6794ba8cd971f023559033

File: internal/controllers/controller_test.go (hunk 22 above)
Comment (lumenalux) [reply internal/controllers/controller_test.go]: Rewritten in 10caf0772fc0da284b2b07b28dff536ac830065b

File: internal/email/message.go (hunk 24 above)
Comment (lumenalux) [reply internal/email/message.go]: Rewritten in 29b26da9664f77ded4e5cc21df5f2b1723c7b08b. Fixed error related to this commit in 4c3ae4f4746fb66cb35ccd1359ccbc543e1df502

File: internal/email/service.go (hunk 27 above)
Comment (lumenalux) [reply internal/email/service.go]: Moved in 7d1d2dcc5224897a8d49c0f1e841349a5df3126f

File: internal/email/service.go (hunk 29 above)
Comment (lumenalux) [reply internal/email/service.go]: Fixed in 087999a6ee5c0c6af1b6a6fabc50a727f9b2a0cf

File: internal/email/service_test.go (hunk 30 above)
Comment (lumenalux) [reply internal/email/service_test.go]: Rewritten in 3134481563b47c9b14cb7b2a3705c62f6d17149c

File: internal/email/send.go (hunk 25 above)
Comment (lumenalux) [reply internal/email/send.go:14]: Renamed in ca6566448de6ccf0706d26c24a23bcb0b56ac22d

File: internal/email/message.go (hunk 23 above)
Comment (lumenalux) [reply internal/email/message.go:31]: I will fix it in HW3

File: .github/workflows/.golangci.yml [hunk 32]
```diff
@@ -0,0 +1,139 @@
+name: golangci
```
Comment (delveper) [inline .github/workflows/.golangci.yml:1]: It would be nice to have relevant file name or separate files for all jobs.

File: pkg/config/config.go [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+	var configuration Config
+
+	data, err := os.ReadFile(filename)
+	if err != nil {
+		return configuration, fmt.Errorf("%w: %s", ErrReadFile, err)
```
[the remark this answers is not in the document]
Comment (lumenalux) [reply pkg/config/config.go]: Wrapped in 821f7b7cd8c652ac9e9efa90e2d7f9ee18126d7e

File: internal/email/send.go (hunk 25 above)
Comment (lumenalux) [reply internal/email/send.go:14]: @delveper This is an interface that is used instead of a specific SMTP client implementation. Standart SMTP client itself has such a method, and it is not the only one. Writing a stub and adapting the project to it is messy

**Review by kachan1208 — APPROVED** [review]
